wandb/integration/kfp/wandb_logging.py

```diff
@@ -0,0 +1,166 @@
+def wandb_log(  # noqa: C901
+    func=None,
+    # /,  # py38 only
+    log_component_file=True,
+):
+    """Wrap a standard python function and log to W&B
+    """
+    import json
+    from functools import wraps
+    from inspect import Parameter, signature
+
+    from kfp import components
+    from kfp.components import (
+        InputArtifact,
+        InputBinaryFile,
+        InputPath,
+        InputTextFile,
+        OutputArtifact,
+        OutputBinaryFile,
+        OutputPath,
+        OutputTextFile,
+    )
+    import wandb
+
+    output_types = (OutputArtifact, OutputBinaryFile, OutputPath, OutputTextFile)
+    input_types = (InputArtifact, InputBinaryFile, InputPath, InputTextFile)
+
+    def isinstance_namedtuple(x):
+        t = type(x)
+        b = t.__bases__
+        if len(b) != 1 or b[0] != tuple:
+            return False
+        f = getattr(t, "_fields", None)
+        if not isinstance(f, tuple):
+            return False
+        return all(type(n) == str for n in f)
+
+    def get_iframe_html(run):
+        return f'<iframe src="{run.url}?kfp=true" style="border:none;width:100%;height:100%;min-width:900px;min-height:600px;"></iframe>'
+
+    def log_input_scalar(name, data, run=None):
+        run.config[name] = data
+        wandb.termlog(f"Setting config: {name} to {data}")
+
+    def log_input_artifact(name, data, type, run=None):
+        artifact = wandb.Artifact(name, type=type)
+        artifact.add_file(data)
+        run.use_artifact(artifact)
+        wandb.termlog(f"Using artifact: {name}")
+
+    def log_output_scalar(name, data, run=None):
+        if isinstance_namedtuple(data):
+            for k, v in zip(data._fields, data):
+                run.log({f"{func.__name__}.{k}": v})
+        else:
+            run.log({name: data})
```

Comment:
I wonder if we should try to align these with an x-axis.  Is this ever called multiple times for the same name?

Reply:
@vanpelt can you clarify what you mean here?  I'm not sure I understand.

Reply:
Calling `run.log({"metric": 0.2})` will associate that data with the default "step" that's incremented every time `run.log` is called.  If in my example, "metric" was actually logged multiple times, you wouldn't be able to line it up with any other metrics in our charts because they would all have different steps.  One approach to getting all of the metrics to lineup is to use `commit=False` which doesn't increment the default "step".  It also doesn't push anything to the server until log is called with `commit=True` which is the default, i.e.

```python
for k, v in zip(data._fields, data):
    run.log({f"{func.__name__}.{k}": v}, commit=False)
run.log({})
```

Would ensure all of the metrics logged in that loop would be comparable at the same "step".  Another approach is to just log the step as a separate metric:

 ```python
 for k, v in zip(data._fields, data):
    run.log({f"{func.__name__}.{k}": v, "global_step": global_step})
 ```
 
 Then you can line those metrics up if you choose "global_step" as your x-axis.